kong/tools/uri.lua

```diff
@@ -12,52 +9,107 @@ local table_concat = table.concat
 local ngx_re_find = ngx.re.find
 local ngx_re_gsub = ngx.re.gsub
 
+local table_new = table.new
+
+
+-- Charset:
+--   reserved = "!" / "*" / "'" / "(" / ")" / ";" / ":" /
+--       "@" / "&" / "=" / "+" / "$" / "," / "/" / "?" / "%" / "#" / "[" / "]"
+--   unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"
+--   other: * (meaning any char that is not mentioned above)
+
+-- Reserved characters have special meaning in URI. Encoding/decoding it affects the semantics of the URI;
+-- Unreserved characters are safe to use as part of HTTP message without encoding;
+-- Other characters has not special meaning but may be not safe to use as part of HTTP message without encoding;
+
+-- We should not unescape or escape reserved characters;
+-- We should unescape but not escape unreserved characters;
+-- We choose to unescape when processing and escape when forwarding for other characters
+
+local RESERVED_CHARS = "!*'();:@&=+$,/?%#[]"
+
+local HYPHEN_BYTE = string_byte('-')
+local DOT_BYTE = string_byte('.')
+local UNDERSCORE_BYTE = string_byte('_')
+local TILDE_BYTE = string_byte('~')
+local CAP_A_BYTE = string_byte('A')
+local CAP_Z_BYTE = string_byte('Z')
+local A_BYTE = string_byte('a')
+local Z_BYTE = string_byte('z')
+local ZERO_BYTE = string_byte('0')
+local NINE_BYTE = string_byte('9')
+
+local CHAR_RESERVED = true
+local CHAR_UNRESERVED = false
+-- nil for CHAR_OTHERS
+
+local chars_to_decode = table_new(256, 0)
```

Comment:
After building this table, we only use it to check for reserved chars (in `normalize_decode`). Seems like maybe this change started out with the intent to use `chars_to_decode` for multiple purposes but wound up limiting its use in the end?

Is there a follow-up changeset/PR that will use `chars_to_decode` and rely on this difference in `CHAR_RESERVED` vs `CHAR_UNRESERVED`? If not, do you think you could clean this up a little so that there's less in here to distract from the actual use-case for this table?

Reply:
We were debating whether to encode `CHAR_OTHERS`. And there was also a solution to allow users to pass in an argument to tweak the behavior. I prefer to at least leave a comment to explain.